File: orion-platform-service/src/services/monitoring/MetricStorageRepository.ts
/**
 * Metric Storage Repository
 *
 * Data access layer for metric_registry and metric_data_points tables.
 * Replaces Map-based storage in MetricCollector with PostgreSQL persistence.
 */

import { DatabasePool } from '../database';
import { MetricQuery } from './MetricCollector';
import { MetricAggregation, DataPoint, MetricSeries } from './types';
import { getCurrentTenantId } from '../../db/tenant-context-storage';

// ==================== Registry Types ====================

export interface MetricRegistryRecord {
  id: string;
  tenant_id: string;
  name: string;
  unit: string;
  default_tags: Record<string, string>;
  description: string | null;
  created_at: Date;
  updated_at: Date;
}

export interface CreateMetricRegistryInput {
  tenant_id?: string;
  name: string;
  unit: string;
  default_tags?: Record<string, string>;
  description?: string;
}

// ==================== Data Point Types ====================

export interface MetricDataPointRecord {
  id: number;
  tenant_id: string;
  metric_name: string;
  value: number;
  tags: Record<string, string>;
  timestamp: Date;
}

export interface InsertDataPointInput {
  tenant_id?: string;
  metric_name: string;
  value: number;
  tags?: Record<string, string>;
  timestamp?: Date;
}

// ==================== Repository Interface ====================

export interface MetricStorageRepository {
  // Registry operations
  registerMetric(input: CreateMetricRegistryInput): Promise<MetricRegistryRecord>;
  unregisterMetric(name: string): Promise<boolean>;
  getAllRegisteredMetrics(): Promise<string[]>;
  getMetricRegistry(name: string): Promise<MetricRegistryRecord | null>;

  // Data point operations
  insertDataPoint(input: InsertDataPointInput): Promise<void>;
  queryMetricSeries(query: MetricQuery, tenantId?: string): Promise<MetricSeries>;
  getLatestValue(name: string, tags?: Record<string, string>, tenantId?: string): Promise<number | null>;

  // Maintenance
  pruneExpired(retentionMs: number, tenantId?: string): Promise<number>;
  clearAll(tenantId?: string): Promise<void>;
}

// ==================== PostgreSQL Implementation ====================

export class PostgresMetricStorageRepository implements MetricStorageRepository {
  constructor(private pool: DatabasePool) {}

  async registerMetric(input: CreateMetricRegistryInput): Promise<MetricRegistryRecord> {
    const tenantId = input.tenant_id || getCurrentTenantId();
    const defaultTags = input.default_tags || {};
    const result = await this.pool.query(
      `INSERT INTO metric_registry (tenant_id, name, unit, default_tags, description)
       VALUES ($1, $2, $3, $4, $5)
       ON CONFLICT (name) DO UPDATE SET
         unit = EXCLUDED.unit,
         default_tags = EXCLUDED.default_tags,
         description = EXCLUDED.description,
         updated_at = NOW()
       RETURNING *`,
      [tenantId, input.name, input.unit, JSON.stringify(defaultTags), input.description || null]
    );
    return this.mapRegistryRow(result.rows[0]);
  }

  async unregisterMetric(name: string): Promise<boolean> {
    const result = await this.pool.query(
      'DELETE FROM metric_registry WHERE name = $1',
      [name]
    );
    return result.rowCount !== null && result.rowCount > 0;
  }

  async getAllRegisteredMetrics(): Promise<string[]> {
    const result = await this.pool.query(
      'SELECT name FROM metric_registry ORDER BY created_at DESC'
    );
    return result.rows.map((row: any) => row.name);
  }

  async getMetricRegistry(name: string): Promise<MetricRegistryRecord | null> {
    const result = await this.pool.query(
      'SELECT * FROM metric_registry WHERE name = $1',
      [name]
    );
    return result.rows.length > 0 ? this.mapRegistryRow(result.rows[0]) : null;
  }

  async insertDataPoint(input: InsertDataPointInput): Promise<void> {
    const tenantId = input.tenant_id || getCurrentTenantId();
    const tags = input.tags || {};
    const timestamp = input.timestamp || new Date();
    await this.pool.query(
      `INSERT INTO metric_data_points (tenant_id, metric_name, value, tags, timestamp)
       VALUES ($1, $2, $3, $4, $5)`,
      [tenantId, input.metric_name, input.value, JSON.stringify(tags), timestamp]
    );
  }

  async queryMetricSeries(query: MetricQuery, tenantId?: string): Promise<MetricSeries> {
    const conditions: string[] = ['metric_name = $1'];
    const params: any[] = [query.name];
    let paramIndex = 2;

    if (tenantId) {
      conditions.push(`tenant_id = $${paramIndex}`);
      params.push(tenantId);
      paramIndex++;
    }

    if (query.startTime) {
      conditions.push(`timestamp >= $${paramIndex}`);
      params.push(query.startTime);
      paramIndex++;
    }

    if (query.endTime) {
      conditions.push(`timestamp <= $${paramIndex}`);
      params.push(query.endTime);
      paramIndex++;
    }

    // Tag filtering using JSONB containment
    if (query.tags && Object.keys(query.tags).length > 0) {
      conditions.push(`tags @> $${paramIndex}`);
      params.push(JSON.stringify(query.tags));
      paramIndex++;
    }

    const whereClause = conditions.join(' AND ');

    const dataResult = await this.pool.query(
      `SELECT value, timestamp FROM metric_data_points
       WHERE ${whereClause}
       ORDER BY timestamp ASC`,
      params
    );

    const points: DataPoint[] = dataResult.rows.map((row: any) => ({
      timestamp: new Date(row.timestamp),
      value: row.value,
    }));

    // Apply max points limit (sampling)
    let sampledPoints = points;
    if (query.maxPoints && points.length > query.maxPoints) {
      const step = Math.ceil(points.length / query.maxPoints);
      sampledPoints = [];
      for (let i = 0; i < points.length && sampledPoints.length < query.maxPoints; i += step) {
        sampledPoints.push(points[i]);
      }
    }

    const values = sampledPoints.map(p => p.value);
    const aggregation = this.computeAggregation(values);

    const windowStart = sampledPoints.length > 0 ? sampledPoints[0].timestamp : new Date();
    const windowEnd = sampledPoints.length > 0 ? sampledPoints[sampledPoints.length - 1].timestamp : new Date();

    return {
      name: query.name,
      dataPoints: sampledPoints,
      aggregation,
      tags: query.tags,
      windowStart,
      windowEnd,
    };
  }

  async getLatestValue(name: string, tags?: Record<string, string>, tenantId?: string): Promise<number | null> {
    const conditions: string[] = ['metric_name = $1'];
    const params: any[] = [name];
    let paramIndex = 2;

    if (tenantId) {
      conditions.push(`tenant_id = $${paramIndex}`);
      params.push(tenantId);
      paramIndex++;
    }

    if (tags && Object.keys(tags).length > 0) {
      conditions.push(`tags @> $${paramIndex}`);
      params.push(JSON.stringify(tags));
      paramIndex++;
    }

    const whereClause = conditions.join(' AND ');

    const result = await this.pool.query(
      `SELECT value FROM metric_data_points
       WHERE ${whereClause}
       ORDER BY timestamp DESC
       LIMIT 1`,
      params
    );

    return result.rows.length > 0 ? result.rows[0].value : null;
  }

  async pruneExpired(retentionMs: number, tenantId?: string): Promise<number> {
    const cutoff = new Date(Date.now() - retentionMs);
    const conditions: string[] = ['timestamp < $1'];
    const params: any[] = [cutoff];

    if (tenantId) {
      conditions.push(`tenant_id = $2`);
      params.push(tenantId);
    }

    const whereClause = conditions.join(' AND ');

    const result = await this.pool.query(
      `DELETE FROM metric_data_points WHERE ${whereClause}`,
      params
    );
    return result.rowCount || 0;
  }

  async clearAll(tenantId?: string): Promise<void> {
    if (tenantId) {
      await this.pool.query('DELETE FROM metric_data_points WHERE tenant_id = $1', [tenantId]);
      await this.pool.query('DELETE FROM metric_registry WHERE tenant_id = $1', [tenantId]);
    } else {
      await this.pool.query('TRUNCATE metric_data_points, metric_registry');
    }
  }

  // ==================== Private Helpers ====================

  private mapRegistryRow(row: any): MetricRegistryRecord {
    return {
      id: row.id,
      tenant_id: row.tenant_id,
      name: row.name,
      unit: row.unit,
      default_tags: row.default_tags || {},
      description: row.description,
      created_at: row.created_at,
      updated_at: row.updated_at,
    };
  }

  private computeAggregation(values: number[]): MetricAggregation {
    if (values.length === 0) {
      return { avg: 0, max: 0, min: 0, p99: 0, p95: 0, count: 0, sum: 0 };
    }

    const sorted = [...values].sort((a, b) => a - b);
    const sum = values.reduce((acc, v) => acc + v, 0);
    const avg = sum / values.length;

    return {
      avg: Math.round(avg * 100) / 100,
      max: sorted[sorted.length - 1],
      min: sorted[0],
      p99: this.percentile(sorted, 99),
      p95: this.percentile(sorted, 95),
      count: values.length,
      sum: Math.round(sum * 100) / 100,
    };
  }

  private percentile(sorted: number[], p: number): number {
    if (sorted.length === 0) return 0;
    if (sorted.length === 1) return sorted[0];

    const index = (p / 100) * (sorted.length - 1);
    const lower = Math.floor(index);
    const upper = Math.ceil(index);

    if (lower === upper) return sorted[lower];

    const weight = index - lower;
    return sorted[lower] * (1 - weight) + sorted[upper] * weight;
  }
}

File: orion-ai-service/src/repositories/metric_storage_repository.py
"""
Metric 仓储层

使用 PostgreSQL 持久化 metric_registry 和 metric_data_points。
对应 TS: src/services/monitoring/MetricStorageRepository.ts
"""

import json
import logging
import os
from contextlib import contextmanager
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

import psycopg2
from psycopg2 import pool

from src.models.metric_models import (
    DataPoint,
    MetricAggregation,
    MetricQuery,
    MetricRegistration,
    MetricSeries,
)

logger = logging.getLogger(__name__)

_DEFAULT_DSN = os.environ.get(
    "ORION_AI_DATABASE_URL",
    "postgresql://orion:orion@localhost:5432/orion",
)


# ==================== 仓储接口 ====================


class MetricStorageRepository:
    """Metric 仓储接口"""

    def register_metric(self, params: MetricRegistration) -> None:
        raise NotImplementedError

    def unregister_metric(self, name: str) -> bool:
        raise NotImplementedError

    def get_all_registered_metrics(self) -> List[str]:
        raise NotImplementedError

    def get_metric_registry(self, name: str) -> Optional[Dict[str, Any]]:
        raise NotImplementedError

    def insert_data_point(
        self,
        tenant_id: str,
        metric_name: str,
        value: float,
        tags: Optional[Dict[str, str]] = None,
        timestamp: Optional[datetime] = None,
    ) -> None:
        raise NotImplementedError

    def query_metric_series(self, query: MetricQuery, tenant_id: str) -> MetricSeries:
        raise NotImplementedError

    def get_latest_value(
        self, name: str, tags: Optional[Dict[str, str]], tenant_id: str
    ) -> Optional[float]:
        raise NotImplementedError

    def prune_expired(self, retention_ms: int, tenant_id: str) -> int:
        raise NotImplementedError

    def clear_all(self, tenant_id: str) -> None:
        raise NotImplementedError


# ==================== PostgreSQL 实现 ====================


class PostgresMetricStorageRepository(MetricStorageRepository):
    """PostgreSQL 实现的 Metric 仓储"""

    def __init__(self, dsn: Optional[str] = None, minconn: int = 1, maxconn: int = 5):
        self._dsn = dsn or _DEFAULT_DSN
        self._pool: Optional[pool.SimpleConnectionPool] = None
        self._minconn = minconn
        self._maxconn = maxconn
        self._tables_ensured = False

    def _get_pool(self) -> pool.SimpleConnectionPool:
        if self._pool is None:
            self._pool = pool.SimpleConnectionPool(
                self._minconn, self._maxconn, self._dsn
            )
        return self._pool

    def _ensure_tables_if_needed(self) -> None:
        if self._tables_ensured:
            return
        conn = self._get_pool().getconn()
        try:
            with conn.cursor() as cur:
                statements = [
                    """
                    CREATE TABLE IF NOT EXISTS metric_registry (
                        id VARCHAR(64) PRIMARY KEY,
                        tenant_id TEXT NOT NULL DEFAULT 'default',
                        name VARCHAR(255) NOT NULL,
                        unit VARCHAR(64) NOT NULL,
                        default_tags JSONB DEFAULT '{}',
                        description TEXT,
                        created_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
                        updated_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
                        CONSTRAINT unique_metric_name_per_tenant UNIQUE (tenant_id, name)
                    )
                    """,
                    """
                    CREATE TABLE IF NOT EXISTS metric_data_points (
                        id SERIAL PRIMARY KEY,
                        tenant_id TEXT NOT NULL DEFAULT 'default',
                        metric_name VARCHAR(255) NOT NULL,
                        value NUMERIC NOT NULL,
                        tags JSONB DEFAULT '{}',
                        timestamp TIMESTAMPTZ NOT NULL DEFAULT NOW()
                    )
                    """,
                    "CREATE INDEX IF NOT EXISTS idx_metric_registry_tenant ON metric_registry(tenant_id)",
                    "CREATE INDEX IF NOT EXISTS idx_metric_registry_name ON metric_registry(name)",
                    "CREATE INDEX IF NOT EXISTS idx_metric_registry_tenant_name ON metric_registry(tenant_id, name)",
                    "CREATE INDEX IF NOT EXISTS idx_metric_data_points_tenant_metric ON metric_data_points(tenant_id, metric_name)",
                    "CREATE INDEX IF NOT EXISTS idx_metric_data_points_timestamp ON metric_data_points(timestamp DESC)",
                    "CREATE INDEX IF NOT EXISTS idx_metric_data_points_metric_timestamp ON metric_data_points(metric_name, timestamp DESC)",
                    "CREATE INDEX IF NOT EXISTS idx_metric_data_points_tags ON metric_data_points USING GIN (tags)",
                ]
                for stmt in statements:
                    cur.execute(stmt)
            conn.commit()
            self._tables_ensured = True
        except Exception as e:
            logger.error(f"Failed to ensure metric tables: {e}")
            conn.rollback()
        finally:
            self._get_pool().putconn(conn)

    @contextmanager
    def _get_conn(self):
        conn = self._get_pool().getconn()
        try:
            yield conn
            conn.commit()
        except Exception:
            conn.rollback()
            raise
        finally:
            self._get_pool().putconn(conn)

    # ==================== Registry 操作 ====================

    def register_metric(self, params: MetricRegistration) -> None:
        """注册指标定义"""
        self._ensure_tables_if_needed()
        tenant_id = params.tenant_id or "default"
        metric_id = f"{tenant_id}:{params.name}"
        default_tags = json.dumps(params.default_tags or {})
        description = params.description or ""
        with self._get_conn() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    """
                    INSERT INTO metric_registry (id, tenant_id, name, unit, default_tags, description)
                    VALUES (%s, %s, %s, %s, %s, %s)
                    ON CONFLICT (tenant_id, name) DO UPDATE SET
                        unit = EXCLUDED.unit,
                        default_tags = EXCLUDED.default_tags,
                        description = EXCLUDED.description,
                        updated_at = NOW()
                    """,
                    (metric_id, tenant_id, params.name, params.unit, default_tags, description),
                )

    def unregister_metric(self, name: str) -> bool:
        """注销指标"""
        self._ensure_tables_if_needed()
        with self._get_conn() as conn:
            with conn.cursor() as cur:
                cur.execute("DELETE FROM metric_registry WHERE name = %s", (name,))
                return cur.rowcount > 0

    def get_all_registered_metrics(self) -> List[str]:
        """获取所有已注册指标名称"""
        self._ensure_tables_if_needed()
        with self._get_conn() as conn:
            with conn.cursor() as cur:
                cur.execute("SELECT name FROM metric_registry ORDER BY created_at DESC")
                return [row[0] for row in cur.fetchall()]

    def get_metric_registry(self, name: str) -> Optional[Dict[str, Any]]:
        """获取指标注册信息"""
        self._ensure_tables_if_needed()
        with self._get_conn() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    "SELECT id, tenant_id, name, unit, default_tags, description, created_at, updated_at FROM metric_registry WHERE name = %s",
                    (name,),
                )
                row = cur.fetchone()
                if row:
                    return {
                        "id": row[0],
                        "tenant_id": row[1],
                        "name": row[2],
                        "unit": row[3],
                        "default_tags": row[4] or {},
                        "description": row[5],
                        "created_at": row[6],
                        "updated_at": row[7],
                    }
                return None

    # ==================== Data Point 操作 ====================

    def insert_data_point(
        self,
        tenant_id: str,
        metric_name: str,
        value: float,
        tags: Optional[Dict[str, str]] = None,
        timestamp: Optional[datetime] = None,
    ) -> None:
        """插入指标数据点"""
        self._ensure_tables_if_needed()
        ts = timestamp or datetime.now(timezone.utc)
        tags_json = json.dumps(tags or {})
        with self._get_conn() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    """
                    INSERT INTO metric_data_points (tenant_id, metric_name, value, tags, timestamp)
                    VALUES (%s, %s, %s, %s, %s)
                    """,
                    (tenant_id, metric_name, value, tags_json, ts),
                )

    def query_metric_series(self, query: MetricQuery, tenant_id: str) -> MetricSeries:
        """查询指标时间序列"""
        self._ensure_tables_if_needed()
        conditions = ["metric_name = %s"]
        params: list = [query.name]
        param_index = 2

        if query.start_time:
            conditions.append(f"timestamp >= %{param_index}")
            params.append(query.start_time)
            param_index += 1
        if query.end_time:
            conditions.append(f"timestamp <= %{param_index}")
            params.append(query.end_time)
            param_index += 1
        if query.tags and query.tags:
            conditions.append(f"tags @> %{param_index}")
            params.append(json.dumps(query.tags))
            param_index += 1

        where_clause = " AND ".join(conditions)
        with self._get_conn() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    f"""
                    SELECT value, timestamp FROM metric_data_points
                    WHERE {where_clause}
                    ORDER BY timestamp ASC
                    """,
                    params,
                )
                rows = cur.fetchall()

        points = [DataPoint(timestamp=row[1], value=float(row[0])) for row in rows]

        # Apply max points limit (sampling)
        if query.max_points and len(points) > query.max_points:
            step = (len(points) + query.max_points - 1) // query.max_points
            points = points[::step][: query.max_points]

        values = [p.value for p in points]
        aggregation = self._compute_aggregation(values)

        window_start = points[0].timestamp if points else datetime.now(timezone.utc)
        window_end = points[-1].timestamp if points else datetime.now(timezone.utc)

        return MetricSeries(
            name=query.name,
            data_points=points,
            aggregation=aggregation,
            tags=query.tags,
            window_start=window_start,
            window_end=window_end,
        )

    def get_latest_value(
        self, name: str, tags: Optional[Dict[str, str]], tenant_id: str
    ) -> Optional[float]:
        """获取指标最新值"""
        self._ensure_tables_if_needed()
        conditions = ["metric_name = %s"]
        params: list = [name]
        param_index = 2

        if tags:
            conditions.append(f"tags @> %{param_index}")
            params.append(json.dumps(tags))
            param_index += 1

        where_clause = " AND ".join(conditions)
        with self._get_conn() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    f"""
                    SELECT value FROM metric_data_points
                    WHERE {where_clause}
                    ORDER BY timestamp DESC
                    LIMIT 1
                    """,
                    params,
                )
                row = cur.fetchone()
                return float(row[0]) if row else None

    def prune_expired(self, retention_ms: int, tenant_id: str) -> int:
        """清理过期的数据点"""
        self._ensure_tables_if_needed()
        cutoff = datetime.now(timezone.utc).timestamp() * 1000 - retention_ms
        cutoff_dt = datetime.fromtimestamp(cutoff / 1000, tz=timezone.utc)
        with self._get_conn() as conn:
            with conn.cursor() as cur:
                cur.execute(
                    "DELETE FROM metric_data_points WHERE tenant_id = %s AND timestamp < %s",
                    (tenant_id, cutoff_dt),
                )
                return cur.rowcount or 0

    def clear_all(self, tenant_id: str) -> None:
        """清空租户的所有指标数据"""
        self._ensure_tables_if_needed()
        with self._get_conn() as conn:
            with conn.cursor() as cur:
                cur.execute("DELETE FROM metric_data_points WHERE tenant_id = %s", (tenant_id,))
                cur.execute("DELETE FROM metric_registry WHERE tenant_id = %s", (tenant_id,))

    # ==================== Private Helpers ====================

    @staticmethod
    def _compute_aggregation(values: List[float]) -> MetricAggregation:
        """计算聚合统计"""
        if not values:
            return MetricAggregation()
        sorted_vals = sorted(values)
        total = sum(values)
        count = len(values)
        avg = total / count
        return MetricAggregation(
            avg=round(avg, 2),
            max=sorted_vals[-1],
            min=sorted_vals[0],
            p99=PostgresMetricStorageRepository._percentile(sorted_vals, 99),
            p95=PostgresMetricStorageRepository._percentile(sorted_vals, 95),
            count=count,
            sum=round(total, 2),
        )

    @staticmethod
    def _percentile(sorted_vals: List[float], p: float) -> float:
        """计算分位值"""
        if not sorted_vals:
            return 0.0
        if len(sorted_vals) == 1:
            return sorted_vals[0]
        index = (p / 100) * (len(sorted_vals) - 1)
        lower = int(index)
        upper = min(lower + 1, len(sorted_vals) - 1)
        if lower == upper:
            return sorted_vals[lower]
        weight = index - lower
        return sorted_vals[lower] * (1 - weight) + sorted_vals[upper] * weight
